File: python/0024. Swap Nodes in Pairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        prev = ListNode(next=head)
        new_head = prev
        node = head

        while node and node.next:
            temp = node.next.next
            node.next.next = node
            prev.next = node.next
            node.next = temp

            prev = node
            node = node.next

        return new_head.next

File: javascript/0024. Swap Nodes in Pairs.js
// INPUT: 1 -> 2 -> 3 -> 4
// OUTPUT: 2 -> 1 -> 4 -> 3

const swapPairs = (head) => {
  let prevNode = null,
    leftNode = head,
    rightNode = head ? head.next : null;

  while (leftNode && rightNode) {
    leftNode.next = rightNode.next;
    rightNode.next = leftNode;
    if (prevNode) {
      prevNode.next = rightNode;
    } else {
      head = rightNode;
    }
    prevNode = leftNode;
    leftNode = leftNode.next;
    rightNode = leftNode ? leftNode.next : null;
  }

  return head;
};

console.log(
  swapPairs({
    val: 1,
    next: { val: 2, next: { val: 3, next: { val: 4, next: null } } },
  })
);
